e2e/nomostest/portforwarder/port_forwarder.go

```diff
@@ -190,67 +196,155 @@ func (pf *PortForwarder) portForwardToPod(pod string, async bool) error {
 	if err != nil {
 		return err
 	}
-	pf.logger.Infof("took %v to wait for port-forward to pod %s/%s (localhost:%d)", took, pf.ns, pod, localPort)
-	pf.setPort(cmd, localPort, pod, async)
+	pf.logger.Infof("took %v to wait for port-forward to pod %s/%s (localhost:%d) (pid:%d)", took, pf.ns, pod, localPort, cmd.Process.Pid)
+	pf.setPort(cmd, localPort, pod)
 	return nil
 }
 
-func (pf *PortForwarder) portForwardToDeployment(async bool) error {
+func (pf *PortForwarder) portForwardToDeployment() error {
 	select {
 	case <-pf.ctx.Done():
 		// Don't bother, subprocess already terminated/terminating
 		return errors.Wrap(pf.ctx.Err(), "failed to start port forward process")
 	default:
 	}
-	// Kill previous port forward process. This isn't strictly necessary since the
-	// subprocess will be killed with the context, and we get a random port every
-	// time. However, this cleans up subprocesses in the interim.
-	if pf.cmd != nil {
-		pf.logger.Infof("stopping port-forward process for %s/%s", pf.ns, pf.deployment)
-		if err := pf.cmd.Process.Kill(); err != nil && !errors.Is(err, os.ErrProcessDone) {
-			return errors.Wrap(err, "failed to kill port forward process")
-		}
-	}
 	pf.logger.Infof("starting port-forward process for %s/%s", pf.ns, pf.deployment)
 	pod, err := pf.kubeClient.GetDeploymentPod(pf.deployment, pf.ns, pf.retryTimeout)
 	if err != nil {
 		return errors.Wrap(err, fmt.Sprintf("failed to get deployment pod for %s/%s", pf.ns, pf.deployment))
 	}
-	err = pf.portForwardToPod(pod.Name, async)
+	err = pf.portForwardToPod(pod.Name)
 	if err != nil {
 		return errors.Wrap(err, fmt.Sprintf("failed to start port forward for %s/%s", pf.ns, pf.deployment))
 	}
 	return nil
 }
 
-// start creates a port forward to a deployment's pod and keeps it open,
+// Start creates a port forward to a deployment's pod and keeps it open,
 // even if the pod is recreated.
-// Once ctx is cancelled, the port forward will be closed and will no longer be recreated
-func (pf *PortForwarder) start() error {
-	if err := pf.portForwardToDeployment(false); err != nil {
-		return err
+// Once ctx is cancelled, the port forward will be closed and will no longer be recreated.
+// Returns a channel. An error will be sent to the channel, and the channel will
+// be closed, when the PortForwarder stops.
+func (pf *PortForwarder) Start() chan error {
+	pf.mux.Lock()
+	defer pf.mux.Unlock()
+	// defend against the PortForwarder being started more than once
+	if pf.isStarted {
```

Comment:
We didn't have to do this, it was a decision I made to change the `PortForwarder` to be scoped to a testcase rather than the entire test suite. It was scoped to the test case originally, and the other `PortForwarder`s are also scoped to the testcase.

It can work either way, but the downside to scoping it to the entire test suite is that if the `PortForwarder` becomes unhealthy it will fail all subsequent tests

Reply:
> if the PortForwarder becomes unhealthy it will fail all subsequent tests

This seems fine. We're making it not flakey :P

Reply:
If we revert this, we also won't be able to consolidate `InitGitRepos` - because that needs to be done in `setupTestCase`

Reply:
Note this is kind of tangential from the `isStarted` logic that I added here. The `isStarated` logic is just to help protect developers from calling `Start` more than once, because that would break the assumption of a single goroutine running for a `PortForwarder`. We would probably still want this regardless of whether we are scoping to a `testcase` or to the whole test suite. The reason this is introduced now is because we are no longer lazily starting, and we've exposed a public `Start` method.

With the logic written as it is now, it wouldn't call `Start` more than once for a `PortForwarder` - but IMO it's good to fail fast if someone accidentally introduces a bug which causes `Start` to be called more than once